exchanges/subscription/subscription.go

```diff
@@ -1,92 +1,128 @@
+	state         State
+	m             sync.RWMutex
 }
 
-// MarshalJSON generates a JSON representation of a Subscription, specifically for config writing
-// The only reason it exists is to avoid having to make Pair a pointer, since that would be generally painful
-// If Pair becomes a pointer, this method is redundant and should be removed
-func (s *Subscription) MarshalJSON() ([]byte, error) {
-	// None of the usual type embedding tricks seem to work for not emitting an nil Pair
-	// The embedded type's Pair always fills the empty value
-	type MaybePair struct {
-		Enabled       bool                   `json:"enabled"`
-		Channel       string                 `json:"channel,omitempty"`
-		Asset         asset.Item             `json:"asset,omitempty"`
-		Params        map[string]interface{} `json:"params,omitempty"`
-		Interval      kline.Interval         `json:"interval,omitempty"`
-		Levels        int                    `json:"levels,omitempty"`
-		Authenticated bool                   `json:"authenticated,omitempty"`
-		Pair          *currency.Pair         `json:"pair,omitempty"`
-	}
+// String implements the Stringer interface for Subscription, giving a human representation of the subscription
+func (s *Subscription) String() string {
+	p := s.Pairs.Format(currency.PairFormat{Uppercase: true, Delimiter: "/"})
+	return fmt.Sprintf("%s %s %s", s.Channel, s.Asset, p.Join())
+}
 
-	k := MaybePair{s.Enabled, s.Channel, s.Asset, s.Params, s.Interval, s.Levels, s.Authenticated, nil}
-	if s.Pair != currency.EMPTYPAIR {
-		k.Pair = &s.Pair
-	}
+// State returns the subscription state
+func (s *Subscription) State() State {
+	s.m.RLock()
+	defer s.m.RUnlock()
+	return s.state
+}
 
-	return json.Marshal(k)
+// SetState sets the subscription state
+// Errors if already in that state or the new state is not valid
+func (s *Subscription) SetState(state State) error {
+	s.m.Lock()
+	defer s.m.Unlock()
+	if state == s.state {
+		return ErrInStateAlready
+	}
+	if state > UnsubscribingState {
+		return ErrInvalidState
+	}
+	s.state = state
+	return nil
 }
 
-// String implements the Stringer interface for Subscription, giving a human representation of the subscription
-func (s *Subscription) String() string {
-	return fmt.Sprintf("%s %s %s", s.Channel, s.Asset, s.Pair)
+// SetKey does what it says on the tin safely for concurrency
+func (s *Subscription) SetKey(key any) {
+	s.m.Lock()
+	defer s.m.Unlock()
+	s.Key = key
 }
 
-// EnsureKeyed sets the default key on a channel if it doesn't have one
-// Returns key for convenience
+// EnsureKeyed returns the subscription key
+// If no key exists then ExactKey will be used
 func (s *Subscription) EnsureKeyed() any {
 	if s.Key == nil {
-		s.Key = DefaultKey{
-			Channel: s.Channel,
-			Asset:   s.Asset,
-			Pair:    s.Pair,
-		}
+		s.Key = &ExactKey{s}
 	}
 	return s.Key
 }
+
+// Clone returns a copy of a subscription
+// Key is set to nil, because any original key is meaningless on a clone
```

Comment:
Most keys contain a pointer to the subscription itself.
If we re-use the same key we'd have a pointer to the original subscription.
But also existentially: The key represents it's identity in the store.
If we Clone the sub then it's because we're going to change it, which means all-bets are off regarding it's identity anyway.

I thought about working out how to re-key it properly, but as it stands now, any `Clone()`ed Sub will get an `ExactKey` when `AddSub`ed, and that suits all use-cases I currently have, and any others would want to SetKey anyway.

Fixed 1fdd21bd
```
// Clone returns a copy of a subscription
// Key is set to nil, because most Key types contain a pointer to the subscription, and because the clone isn't added to the store yet
// Users should allow a default key to be assigned on AddSubscription or can SetKey as necessary
```
Hope that works for you.